### DISCUSSION TITLE
How to use DPC++ compiler for CPU with JIT support and SIMD code generation ?

### DISCUSSION THREAD

**Discussion by lohika-denis-kotov** [intel/llvm#6998] [General]
I have a question how to use DPC++ compiler for CPU with JIT support and SIMD code generation ?
It is need for some of the projects, for example https://github.com/openvinotoolkit/openvino/tree/master/src/plugins/intel_cpu plugin ...

I have not found the option for DPC++ compiler for such feature.
Closest what I have found is https://www.intel.com/content/www/us/en/develop/documentation/oneapi-dpcpp-cpp-compiler-dev-guide-and-reference/top/compiler-reference/compiler-options/code-generation-options/arch.html#arch, but this option is only available for Windows !!

Have somebody know how to do it ?

**Comment by lohika-denis-kotov** [comment]
One comment on technical side how it could be done.
On code generation side the following code could be compiled for first 2 dimensions could be parallelized using `oneTBB` library and last dimension could be parallelized using CPU SIMD instructions (SSE4.1, AVX2, AVX512 and etc.):
```cpp
        cgh.parallel_for<class Fillbuffer>(
            accessor.get_range(), [=](cl::sycl::id<1> WIid1, cl::sycl::id<2> WIid2, cl::sycl::id<3> WIid3) {
                // Fill buffer with indexes
                
            });
```

**Reply by bader** [reply]
What you described is exactly how DPC++ program works on CPU when executed via OpenCL backend. This is default behavior, so noting special is needed. @lohika-denis-kotov, if you observe different behavior, could you provide a reproducer, so we investigate it, please?
I also should mention that OpenCL CPU JIT compiler has auto-vectorization module, which might keep code scalar (i.e. not utilize SIMD instructions) if it considers scalar code to be faster to execute.

**Reply by lohika-denis-kotov** [reply]
@bader 
> What you described is exactly how DPC++ program works on CPU when executed via OpenCL backend. This is default behavior, so noting special is needed. @lohika-denis-kotov, if you observe different behavior, could you provide a reproducer, so we investigate it, please? I also should mention that OpenCL CPU JIT compiler has auto-vectorization module, which might keep code scalar (i.e. not utilize SIMD instructions) if it considers scalar code to be faster to execute.

Do you mean that for OpenCL backend `oneTBB` is used for CPU device ?
Also what about JIT ? According the documentation for me looks like for CPU AOT is used by default ... Am I wrong ?

**Reply by bader** [reply]
OpenCL CPU implementation uses oneTBB to execute kernels in parallel.
JIT compilation is used unless you explicitly specifying CPU target triple, which is not a default option.

**Reply by lohika-denis-kotov** [reply]
> OpenCL CPU implementation uses oneTBB to execute kernels in parallel. JIT compilation is used unless you explicitly specifying CPU target triple, which is not a default option.

@bader Is there a way to dump what SIMD instructions are generated during JIT phase ? It is sometimes needed for optimization

**Reply by bader** [reply]
Please, ask this question at https://software.intel.com/en-us/forums/opencl forum. JIT phase is performed by OpenCL CPU and my knowledge most likely is not relevant for recent versions.

**Reply by lohika-denis-kotov** [reply]
@bader Okay, thanks !!

**Reply by lohika-denis-kotov** [reply]
> OpenCL CPU implementation uses oneTBB to execute kernels in parallel. JIT compilation is used unless you explicitly specifying CPU target triple, which is not a default option.

@bader I've found that this statement contradict https://intel.github.io/llvm-docs/GetStartedGuide.html#cpu:
```
CPU AOT compiler opencl-aot is enabled by default ...
```

**Reply by bader** [reply]
No contradiction. This statement says that `the build of opencl-aot tool is enabled by default`. It's not the same as "DPC++ compiler uses AOT compilation mode for CPU device by default".

**Reply by lohika-denis-kotov** [reply]
Thanks, it is my misunderstood
